pkg/storage/iterator.go

```diff
@@ -0,0 +1,386 @@
+}
+
+func buildIterators(ctx context.Context, chks map[model.Fingerprint][][]*chunkenc.LazyChunk, filter logql.Filter, direction logproto.Direction, from, through time.Time) ([]iter.EntryIterator, error) {
+	result := make([]iter.EntryIterator, 0, len(chks))
+	for _, chunks := range chks {
+		iterator, err := buildHeapIterator(ctx, chunks, filter, direction, from, through)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, iterator)
+	}
+
+	return result, nil
+}
+
+func buildHeapIterator(ctx context.Context, chks [][]*chunkenc.LazyChunk, filter logql.Filter, direction logproto.Direction, from, through time.Time) (iter.EntryIterator, error) {
+	result := make([]iter.EntryIterator, 0, len(chks))
+	if chks[0][0].Chunk.Metric.Has("__name__") {
+		labelsBuilder := labels.NewBuilder(chks[0][0].Chunk.Metric)
+		labelsBuilder.Del("__name__")
+		chks[0][0].Chunk.Metric = labelsBuilder.Labels()
+	}
+	labels := chks[0][0].Chunk.Metric.String()
+
+	for i := range chks {
+		iterators := make([]iter.EntryIterator, 0, len(chks[i]))
+		for j := range chks[i] {
+			iterator, err := chks[i][j].Iterator(ctx, from, through, direction, filter)
+			if err != nil {
+				return nil, err
+			}
+			iterators = append(iterators, iterator)
+		}
+		if direction == logproto.BACKWARD {
+			for i, j := 0, len(iterators)-1; i < j; i, j = i+1, j-1 {
+				iterators[i], iterators[j] = iterators[j], iterators[i]
+			}
+		}
+		result = append(result, iter.NewNonOverlappingIterator(iterators, labels))
+	}
+
+	return iter.NewHeapIterator(result, direction), nil
+}
+
+func filterSeriesByMatchers(chks map[model.Fingerprint][][]*chunkenc.LazyChunk, matchers []*labels.Matcher) map[model.Fingerprint][][]*chunkenc.LazyChunk {
+outer:
+	for fp, chunks := range chks {
+		for _, matcher := range matchers {
+			if !matcher.Matches(chunks[0][0].Chunk.Metric.Get(matcher.Name)) {
+				delete(chks, fp)
+				continue outer
+			}
+		}
+	}
+	return chks
+}
+
+func fetchLazyChunks(ctx context.Context, chunks []*chunkenc.LazyChunk) error {
+	log, ctx := spanlogger.New(ctx, "LokiStore.fetchLazyChunks")
+	defer log.Finish()
+
+	var totalChunks int
+	chksByFetcher := map[*chunk.Fetcher][]*chunkenc.LazyChunk{}
+	for _, c := range chunks {
+		if c.Chunk.Data == nil {
+			chksByFetcher[c.Fetcher] = append(chksByFetcher[c.Fetcher], c)
+			totalChunks++
+		}
+	}
+	if len(chksByFetcher) == 0 {
+		return nil
+	}
+	level.Debug(log).Log("msg", "loading lazy chunks", "chunks", totalChunks)
+
+	errChan := make(chan error)
+	for fetcher, chunks := range chksByFetcher {
+		go func(fetcher *chunk.Fetcher, chunks []*chunkenc.LazyChunk) {
+
+			keys := make([]string, 0, len(chunks))
+			chks := make([]chunk.Chunk, 0, len(chunks))
+			index := make(map[string]*chunkenc.LazyChunk, len(chunks))
+
+			// FetchChunks requires chunks to be ordered by external key.
+			sort.Slice(chunks, func(i, j int) bool { return chunks[i].Chunk.ExternalKey() < chunks[j].Chunk.ExternalKey() })
```

Comment:
@gouthamve we talked about this, hopefully this will resolve the warning about `got chunk from cache we didn't ask for`